### DISCUSSION TITLE
Colour cycle on path segment

### DISCUSSION THREAD

**Discussion by quiltedstars** [JanDeDobbeleer/oh-my-posh#3751] [Troubleshoot] [answered]
### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### What happened?

If this is possible, you have provided me everything I've ever hoped for...
Noticed that cycling colours is now a thing. Manual says that you can use it on the path segment under properties: "a list of foreground colors to cycle through to colorize the individual path folders". I take this to mean that, for example, the home folder can be one colour, the next folder can be another, and so on. Firstly, is that the intention?
If so, I can't figure out how to get it to work. Currently, my segment looks like this:
```
        {
  "type": "path",
  "style": "powerline",
  "powerline_symbol": " ",
  "foreground": "#ffffff",
  "template": "{{ .Path }} ",
    "background": "#4D6206",  
    "cycle": [
  {
    "background": "#4D6206",
    "foreground": "#ffffff"
  },
  {
    "background": "#AA6FFB",
    "foreground": "#ffffff"
  }
],        
  "properties": {
    "style": "full",
            "max_depth": 100,
                "home_icon": "\uf7db",
                "folder_separator_icon": " <transparent>\ueab6</> ",
    "mapped_locations": {
      "C:\\temp": ""
    }
  }
}
```
I have tried putting the cycle section under properties, along with omitting the background colour variable, but doing the latter makes the prompt glitch out, and putting cycle under properties changes nothing (as it currently does now). I am just wondering how to get it to work.

### Theme

Agnoster minimal.

### What OS are you seeing the problem on?

Linux

### Which shell are you using?

zsh

### Log output

```Shell
Version: 14.32.2

Shell: zsh (5.9)

Segments:

ConsoleTitle(false)                       -   0 ms - 
session(true)                             -   0 ms -    
path(true)                                -   0 ms -    Documents 

Run duration: 3.033572ms

Cache path: /home/mags2/.cache/oh-my-posh

Config path: /home/mags2/.poshthemes/agnoster.minimal.omp.json

Logs:

[DEBUG] 23:26:04.924 shell.go:Getenv:398 → NO DATA
[TRACE] 23:26:04.924 shell.go:Getenv(TERM_PROGRAM) - 10.921µs
[DEBUG] 23:26:04.924 shell.go:Getenv:398 → NO DATA
[TRACE] 23:26:04.924 shell.go:Getenv(XDG_CACHE_HOME) - 1.971µs
[TRACE] 23:26:04.924 shell_unix.go:CachePath() - 11.618µs
[DEBUG] 23:26:04.926 shell.go:Getenv:398 → /home/mags2/.poshthemes/agnoster.minimal.omp.json
[TRACE] 23:26:04.926 shell.go:Getenv(POSH_THEME) - 6.059µs
[DEBUG] 23:26:04.926 shell_unix.go:Platform:71 → void
[TRACE] 23:26:04.926 shell.go:resolveConfigPath() - 15.214µs
[TRACE] 23:26:04.926 shell.go:Init() - 1.763583ms
[TRACE] 23:26:04.926 shell.go:Flags() - 48ns
[TRACE] 23:26:04.926 config.go:loadConfig() - 154.829µs
[TRACE] 23:26:04.926 shell.go:Flags() - 40ns
[DEBUG] 23:26:04.926 shell.go:Getenv:398 → NO DATA
[TRACE] 23:26:04.926 shell.go:Getenv(OMP_CACHE_DISABLED) - 1.807µs
[DEBUG] 23:26:04.927 shell.go:Shell:645 → process name: zsh
[TRACE] 23:26:04.927 shell.go:Shell() - 504.084µs
[DEBUG] 23:26:04.927 shell.go:Getenv:398 → 5.9
[TRACE] 23:26:04.927 shell.go:Getenv(POSH_SHELL_VERSION) - 2.192µs
[DEBUG] 23:26:04.927 debug.go:PrintDebug:22 → Segment: Title
[TRACE] 23:26:04.927 shell.go:Flags() - 46ns
[DEBUG] 23:26:04.927 block.go:Debug:230 → Segment: session
[TRACE] 23:26:04.927 shell.go:GOOS() - 50ns
[DEBUG] 23:26:04.927 shell.go:Pwd:428 → /home/mags2/Documents
[TRACE] 23:26:04.927 shell.go:Pwd() - 6.83µs
[TRACE] 23:26:04.927 shell.go:GOOS() - 23ns
[DEBUG] 23:26:04.927 shell.go:Getenv:398 → NO DATA
[TRACE] 23:26:04.927 shell.go:Getenv(SSH_CONNECTION) - 1.702µs
[DEBUG] 23:26:04.927 shell.go:Getenv:398 → NO DATA
[TRACE] 23:26:04.927 shell.go:Getenv(SSH_CLIENT) - 1.853µs
[TRACE] 23:26:04.927 shell_unix.go:Root() - 265ns
[TRACE] 23:26:04.927 shell.go:Shell() - 53ns
[TRACE] 23:26:04.927 shell.go:ErrorCode() - 49ns
[DEBUG] 23:26:04.927 shell_unix.go:IsWsl:35 → false
[TRACE] 23:26:04.927 shell_unix.go:IsWsl() - 2.575µs
[TRACE] 23:26:04.927 shell.go:Pwd() - 70ns
[TRACE] 23:26:04.927 shell.go:GOOS() - 21ns
[TRACE] 23:26:04.927 shell.go:GOOS() - 21ns
[TRACE] 23:26:04.927 shell.go:GOOS() - 22ns
[TRACE] 23:26:04.927 shell.go:GOOS() - 23ns
[TRACE] 23:26:04.927 shell.go:GOOS() - 23ns
[TRACE] 23:26:04.927 shell.go:GOOS() - 23ns
[TRACE] 23:26:04.927 shell.go:GOOS() - 22ns
[TRACE] 23:26:04.927 shell.go:GOOS() - 22ns
[TRACE] 23:26:04.927 shell.go:GOOS() - 23ns
[TRACE] 23:26:04.927 shell.go:GOOS() - 22ns
[TRACE] 23:26:04.927 shell.go:GOOS() - 20ns
[DEBUG] 23:26:04.927 shell.go:User:549 → mags2
[TRACE] 23:26:04.927 shell.go:User() - 2.006µs
[DEBUG] 23:26:04.927 shell.go:Host:561 → mags
[TRACE] 23:26:04.927 shell.go:Host() - 2.473µs
[TRACE] 23:26:04.927 shell.go:GOOS() - 22ns
[DEBUG] 23:26:04.927 shell_unix.go:Platform:71 → void
[DEBUG] 23:26:04.927 shell.go:Getenv:398 → 1
[TRACE] 23:26:04.927 shell.go:Getenv(SHLVL) - 1.672µs
[TRACE] 23:26:04.927 shell.go:TemplateCache() - 74.148µs
[TRACE] 23:26:04.927 shell.go:Shell() - 46ns
[DEBUG] 23:26:04.927 block.go:Debug:230 → Segment: path
[TRACE] 23:26:04.927 shell.go:Pwd() - 32ns
[TRACE] 23:26:04.927 shell.go:GOOS() - 43ns
[TRACE] 23:26:04.927 shell.go:Pwd() - 42ns
[TRACE] 23:26:04.927 shell.go:Shell() - 24ns
[TRACE] 23:26:04.927 shell.go:Shell() - 25ns
[DEBUG] 23:26:04.927 properties.go:GetBool:22 → mapped_locations_enabled: true
[DEBUG] 23:26:04.927 properties.go:GetString:28 → 
[DEBUG] 23:26:04.927 properties.go:GetString:28 → 
[TRACE] 23:26:04.927 shell.go:GOOS() - 20ns
[DEBUG] 23:26:04.927 properties.go:GetString:28 → 
[DEBUG] 23:26:04.927 properties.go:GetKeyValueMap:46 → mapped_locations: map[C:\temp:]
[TRACE] 23:26:04.927 shell.go:GOOS() - 19ns
[TRACE] 23:26:04.927 shell.go:PathSeparator() - 35ns
[TRACE] 23:26:04.927 shell.go:GOOS() - 20ns
[TRACE] 23:26:04.927 shell.go:GOOS() - 21ns
[TRACE] 23:26:04.927 shell.go:GOOS() - 35ns
[TRACE] 23:26:04.927 shell.go:GOOS() - 21ns
[TRACE] 23:26:04.927 shell.go:PathSeparator() - 20ns
[TRACE] 23:26:04.927 shell.go:PathSeparator() - 21ns
[TRACE] 23:26:04.927 shell.go:GOOS() - 20ns
[TRACE] 23:26:04.927 shell.go:GOOS() - 19ns
[TRACE] 23:26:04.927 shell.go:PathSeparator() - 20ns
[TRACE] 23:26:04.927 shell.go:GOOS() - 21ns
[TRACE] 23:26:04.927 shell.go:GOOS() - 20ns
[TRACE] 23:26:04.927 shell.go:PathSeparator() - 20ns
[TRACE] 23:26:04.927 shell.go:GOOS() - 18ns
[TRACE] 23:26:04.927 shell.go:GOOS() - 19ns
[TRACE] 23:26:04.927 shell.go:PathSeparator() - 20ns
[TRACE] 23:26:04.927 shell.go:GOOS() - 21ns
[TRACE] 23:26:04.927 shell.go:GOOS() - 20ns
[TRACE] 23:26:04.927 shell.go:PathSeparator() - 20ns
[TRACE] 23:26:04.927 shell.go:PathSeparator() - 19ns
[DEBUG] 23:26:04.927 properties.go:GetString:28 → full
[TRACE] 23:26:04.927 shell.go:PathSeparator() - 21ns
[DEBUG] 23:26:04.927 properties.go:GetStringArray:52 → cycle: []
[DEBUG] 23:26:04.927 properties.go:GetString:28 → NO DATA
[TRACE] 23:26:04.927 shell.go:PathSeparator() - 21ns
[DEBUG] 23:26:04.927 properties.go:GetString:28 →  <transparent></> 
[TRACE] 23:26:04.927 shell.go:PathSeparator() - 20ns
[TRACE] 23:26:04.927 shell.go:Pwd() - 46ns
[DEBUG] 23:26:04.927 shell_unix.go:IsWsl:35 → false
[TRACE] 23:26:04.927 shell_unix.go:IsWsl() - 1.765µs
[TRACE] 23:26:04.927 shell.go:StackCount() - 55ns
[TRACE] 23:26:04.927 shell_unix.go:DirIsWritable(/home/mags2/Documents) - 2.258µs
[TRACE] 23:26:04.927 shell.go:TemplateCache() - 45ns
[TRACE] 23:26:04.927 shell.go:TemplateCache() - 46ns
[TRACE] 23:26:04.927 shell.go:Shell() - 38ns
[DEBUG] 23:26:04.927 shell.go:Getenv:398 → NO DATA
[TRACE] 23:26:04.927 shell.go:Getenv(XDG_CACHE_HOME) - 7.922µs
[TRACE] 23:26:04.927 shell_unix.go:CachePath() - 16.014µs
[TRACE] 23:26:04.927 shell.go:Flags() - 58ns
```

**Comment by JanDeDobbeleer** [comment]
@quiltedstars cycle is `[]string`, so:

```json
{
  "type": "path",
  "style": "powerline",
  "powerline_symbol": " ",
  "foreground": "#ffffff",
  "template": "{{ .Path }} ",
  "background": "#4D6206",
  "properties": {
    "style": "full",
    "max_depth": 100,
    "home_icon": "\uf7db",
    "folder_separator_icon": " <transparent>\ueab6</> ",
    "mapped_locations": {
      "C:\\temp": ""
    },
    "cycle": [
      "#ffffff,#4D6206",
      "#ffffff,#AA6FFB"
    ]
  }
}
```

**Reply by quiltedstars** [reply]
Ah brilliant! Thank you for such an expedient reply. I would guess there's no way currently to have the colour change applied to the background of the folder separator icon? As it is now, the coloured paths only colour the actual path name, and doesn't extend to the individual path 'segment'. Here's a photo: https://imgur.com/VK4tpRB @JanDeDobbeleer

**Reply by JanDeDobbeleer** [reply, marked as answer]
@quiltedstars ah, yes. No, that's not how this works. But maybe that could be an enhancement. I'll try to have a look this week.

**Reply by quiltedstars** [reply]
@JanDeDobbeleer Thank you! I imagine that because cycling can apply to the prompt and a specific segment, that perhaps there's something in the code that could apply the cycling specifically to the folder separator icons, though that may be complicated by the fact that icons like an arrow can't be divided neatly in half.

**Reply by JanDeDobbeleer** [reply]
@quiltedstars #3760

**Reply by quiltedstars** [reply]
@JanDeDobbeleer Oh my goodness, thank you so so much! I hate to be more of a bother considering you are were already kind enough to take this on, but is there any way to get the folder icon separator to be coloured (and subsequently cycled) according to the background and not the foreground color? Playing around, I've got this setup now, and it would be complete if I could have the icon color be the preceding background color. If this isn't possible or too much work (since it currently takes on the foreground color of the element _after_ it, and I imagine that might cause some issues because I imagine the foreground colour is tied to how the path reads each file (ie. /home/user is, necessarily, composed of /home and /user, not home/ and user/)) then no worries at all. You've already gone above and beyond what I could have expected!

https://imgur.com/O4sEatL

EDIT: Realize I can solve this (in my use case of only cycling twice) by having alternating colours, ie. "#1e1e1e,#ffffff", "#ffffff,#1e1e1e", but this wouldn't work if someone used more than two cycles

**Reply by JanDeDobbeleer** [reply]
@quiltedstars that's a really difficult one as it resets the colors per path folder. I gave it a go but it's not something we can add unless we also add a pointer to the previously used color (e.g. `previousBackground`/`previousForeground`).

**Reply by MeRayORG** [reply]
> https://imgur.com/O4sEatL

Sorry for necroing a two-year-old post, but could you maybe share how you got that done? I'm dealing with the reverse issue right now, I can style the separator perfectly fine, but I can't change the foreground to the background color of each folder.

![image](https://github.com/user-attachments/assets/7d81dab0-8ad5-4260-b0e2-a8ba7a0e8e0d)

And since we highly likely want exactly the same thing, did you ever get it to work?

**Reply by JanDeDobbeleer** [reply]
can you highlight what you want to see changed here @MeRayORG ?

**Reply by MeRayORG** [reply]
Thanks for responding so quickly absolutely awesome :smile:.
Ultimately I'd like to have the folders look like their own segments (with unique background color) and the same for the next folder separator. Maybe an easily extendable configuration would have the current cycle color made available to the folder_format and folder_separator_icon with something like .CycleColor if that's possible?
